src/internal/storage/fileset/index/writer.go

```diff
@@ -121,43 +130,45 @@ func (w *Writer) callback(level int) chunk.WriterCallback {
 			LastPath: lastPath,
 			ChunkRef: chunk.Reference(dataRef),
 		}
-		// Set the root index when the writer is closed and we are at the top index level.
-		if w.closed {
-			w.root = idx
-		}
-		// Create next index level if it does not exist.
-		if level == len(w.levels)-1 {
-			cw := w.chunks.NewWriter(w.ctx, uuid.NewWithoutDashes(), w.callback(level+1), chunk.WithRollingHashConfig(averageBits, int64(level+1)))
-			w.levels = append(w.levels, &levelWriter{
-				cw:  cw,
-				pbw: pbutil.NewWriter(cw),
-			})
-		}
 		// Write index entry in next index level.
 		return w.writeIndex(idx, level+1)
 	}
 }
 
 // Close finishes the index, and returns the serialized top index level.
 func (w *Writer) Close() (ret *Index, retErr error) {
-	w.mu.Lock()
-	w.closed = true
-	w.mu.Unlock()
-
 	// Note: new levels can be created while closing, so the number of iterations
 	// necessary can increase as the levels are being closed. Levels stop getting
-	// created when the top level chunk writer has been closed and the number of
-	// annotations and chunks it has is one (one annotation in one chunk).
-	for i := 0; i < len(w.levels); i++ {
-		w.mu.Lock()
-		l := w.levels[i]
-		w.mu.Unlock()
+	// created when the top level chunk writer has been closed and the level chunk
+	// writer's first and last index are the same.
+	// Duplicate indeces may occur if a split point is found in a written index.
+	for i := 0; i < w.numLevels(); i++ {
+		l := w.getLevel(i)
 		if err := l.cw.Close(); err != nil {
 			return nil, err
 		}
-		if l.cw.AnnotationCount() == 1 && l.cw.ChunkCount() == 1 {
-			break
+		if l.firstIdx.Path == l.lastIdx.Path {
```

Comment:
Also, could we rename the channel to `done`. That is a common name for a channel being used like this.

Reply:
why unbuffered channel? Wouldn't that force us to block until the callback gets called?

Reply:
I was referring to the pattern here: https://gobyexample.com/closing-channels to prevent multiple receives from being problematic. I actually think we would be better off with a cancel context for the index writer and doing a defer cancel in the close. This would error out the leftover writers and not require additional channel management.